core/iwasm/libraries/lib-wasi-threads/stress_test/mutex_common.h

```diff
@@ -0,0 +1,261 @@
+        if (status_code == EAGAIN) {
+            usleep(timeout_us);
+            timeout_us *= 2;
+        }
+    }
+
+    assert(status_code == 0 && "Thread creation should succeed");
+}
+
+// We're counting how many times each thread was called using this array
+// Main thread is also counted here so we need to make arrays bigger
+typedef struct {
+    int tids[NUM_THREADS + 1];
+    int calls[NUM_THREADS + 1];
+} StatCollector;
+
+StatCollector call_stat;
+
+void
+clear_stat()
+{
+    for (int i = 0; i < NUM_THREADS + 1; ++i) {
+        call_stat.tids[i] = 0;
+        call_stat.calls[i] = 0;
+    }
+}
+
+void
+add_to_stat(int tid)
+{
+    int tid_num = 0;
+    for (; tid_num < NUM_THREADS + 1 && call_stat.tids[tid_num] != 0;
+         ++tid_num) {
+        if (call_stat.tids[tid_num] == tid) {
+            call_stat.calls[tid_num]++;
+            return;
+        }
+    }
+
+    assert(tid_num < NUM_THREADS + 1);
+    call_stat.tids[tid_num] = tid;
+    call_stat.calls[tid_num] = 1;
+}
+
+// This function ensures that every of NUM_THREADS+1 threads number of
+// operations is in between [NUM_ITER/NUM_THREADS, NUM_ITER/(NUM_THREADS+2)].
+// For 100'000 operations it means that every thread should be called [7142,
+// 8333] times. Also, such a simple hash function might be not the ideal one so
+// the confidence interval is kept quite wide
+void
+check_stat()
+{
+    for (int i = 0; i < NUM_THREADS + 1; ++i) {
+        if (call_stat.calls[i] != 0) {
+            assert(call_stat.calls[i] < NUM_ITER / NUM_THREADS
+                   && "Call distribution by TID is out of range");
+            assert(call_stat.calls[i] > NUM_ITER / (NUM_THREADS + 2)
+                   && "Call distribution by TID is out of range");
+        }
+    }
+}
+
+void
+print_stat()
+{
+    fprintf(stderr, "Thread calls count by TID\n");
+    for (int i = 0; i < NUM_THREADS + 1; ++i) {
+        if (call_stat.tids[i] != 0) {
+            fprintf(stderr, "TID: %d; Calls: %d\n", call_stat.tids[i],
+                    call_stat.calls[i]);
+        }
+    }
+}
+
+// xor with a random value usually gives quite a uniform distribution
+uint32_t
+hash(uint32_t counter)
+{
+    return (counter << 22) ^ (0xA2BC22A); // Random value for hash func
+}
+
+// Locking and unlocking the same thread
+void
+same_thread_test(pthread_mutex_t *mutex)
+{
+    assert(pthread_mutex_lock(mutex) == 0
+           && "Main thread should be able to lock a mutex");
+    assert(pthread_mutex_unlock(mutex) == 0
+           && "Main thread should be able to unlock a mutex");
+}
+
+void *
+inc_until_limit(void *arg)
+{
+    MutexCounter mutex_counter = *(MutexCounter *)(arg);
+    int sleep_us = 0;
+    while (!pthread_mutex_lock(mutex_counter.mutex)
+           && *mutex_counter.counter < mutex_counter.limit) {
+        (*mutex_counter.counter)++;
+        add_to_stat((int)(pthread_self()));
+        if (mutex_counter.is_sleeping) {
+            sleep_us = hash(*mutex_counter.counter) % 1000;
```

Comment:
why do we hash the counter? Is it to generate a random value? If so, why not to use rand() instead?

Reply:
Also, shouldn't this assignment be moved to L151?

Reply:
done

Reply:
>Also, shouldn't this assignment be moved to L151?

both rand() and hash() that I used before are **not** thread-safe, that's why it's inside the mutex